Crash

Running MyPy 0.950 on [the latest efax 1.6.3](https://github.com/NeilGirdhar/efax) gives
```shell
❯ mypy efax --show-traceback
efax/_src/distributions/dirichlet.py:29: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.950
Traceback (most recent call last):
  File "mypy/semanal.py", line 5301, in accept
  File "mypy/nodes.py", line 1021, in accept
  File "mypy/semanal.py", line 1121, in visit_class_def
  File "mypy/semanal.py", line 1201, in analyze_class
  File "mypy/semanal.py", line 1210, in analyze_class_body_common
  File "mypy/semanal.py", line 1256, in apply_class_plugin_hooks
  File "mypy/plugins/dataclasses.py", line 524, in dataclass_class_maker_callback
  File "mypy/plugins/dataclasses.py", line 120, in transform
  File "mypy/plugins/dataclasses.py", line 408, in collect_attributes
  File "mypy/plugins/dataclasses.py", line 106, in expand_typevar_from_subtype
  File "mypy/typeops.py", line 151, in map_type_from_supertype
  File "mypy/maptype.py", line 24, in map_instance_to_supertype
  File "mypy/maptype.py", line 37, in map_instance_to_supertypes
  File "mypy/maptype.py", line 82, in map_instance_to_direct_supertypes
  File "mypy/expandtype.py", line 18, in expand_type
  File "mypy/types.py", line 1127, in accept
  File "mypy/expandtype.py", line 90, in visit_instance
  File "mypy/expandtype.py", line 212, in expand_types
  File "mypy/types.py", line 2472, in accept
AssertionError: 
efax/_src/distributions/dirichlet.py:29: : note: use --pdb to drop into pdb
```

You can also repro it with just one class, which is probably still tickling the same bug:
```python
import dataclasses
import typing

T = typing.TypeVar('T', bound='Foo')

@dataclasses.dataclass
class Foo(typing.Generic[T]):
	data: int
```

If you remove either of the following elements, the problem won't be reproducible:

* The `data` field.
* The bound on `T`.

The bound need not be `Foo` itself, as long as it is a *necessary* forward reference - that is, a forward reference to a class defined below T (and therefore "necessary" because, if you removed the quotes, it would fail with a NameError). If you put e.g. `'object'`, or the name of a class which already exists, then it won't reproduce. But you can just put something like `class Bar: pass` below the declaration of T, and then `bound='Bar'` will reproduce the bug.

The traceback is also slightly different:
```
example.py:8: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.960+dev.612074f05ba5d65e016c031dc94b13f962d9c6b4
Traceback (most recent call last):
  File "C:\Users\Kevin\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\Kevin\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\Kevin\poker\mypy-testing\Scripts\mypy.exe\__main__.py", line 7, in <module>
    sys.exit(console_entry())
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\__main__.py", line 12, in console_entry
    main(None, sys.stdout, sys.stderr)
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\main.py", line 96, in main
    res, messages, blockers = run_build(sources, options, fscache, t0, stdout, stderr)
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\main.py", line 173, in run_build
    res = build.build(sources, options, None, flush_errors, fscache, stdout, stderr)
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\build.py", line 181, in build
    result = _build(
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\build.py", line 257, in _build
    graph = dispatch(sources, manager, stdout)
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\build.py", line 2752, in dispatch
    process_graph(graph, manager)
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\build.py", line 3110, in process_graph
    process_stale_scc(graph, scc, manager)
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\build.py", line 3202, in process_stale_scc
    mypy.semanal_main.semantic_analysis_for_scc(graph, scc, manager.errors)
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\semanal_main.py", line 79, in semantic_analysis_for_scc
    process_functions(graph, scc, patches)
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\semanal_main.py", line 233, in process_functions
    process_top_level_function(analyzer,
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\semanal_main.py", line 274, in process_top_level_function
    deferred, incomplete, progress = semantic_analyze_target(target, state, node, active_type,
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\semanal_main.py", line 326, in semantic_analyze_target
    analyzer.refresh_partial(refresh_node,
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\semanal.py", line 417, in refresh_partial
    self.accept(node)
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\semanal.py", line 5344, in accept
    node.accept(self)
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\nodes.py", line 747, in accept
    return visitor.visit_func_def(self)
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\semanal.py", line 637, in visit_func_def
    self.analyze_func_def(defn)
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\semanal.py", line 669, in analyze_func_def
    self.defer(defn)
  File "C:\Users\Kevin\poker\mypy-testing\lib\site-packages\mypy\semanal.py", line 5044, in defer
    assert not self.final_iteration, 'Must not defer during final iteration'
AssertionError: Must not defer during final iteration
example.py:8: : note: use --pdb to drop into pdb
```

--- a/mypy/plugin.py
+++ b/mypy/plugin.py
@@ -692,7 +692,8 @@ class Plugin(CommonPluginApi):
 
         The plugin can modify a TypeInfo _in place_ (for example add some generated
         methods to the symbol table). This hook is called after the class body was
-        semantically analyzed, but *there may still be placeholders*.
+        semantically analyzed, but *there may still be placeholders* (typically
+        caused by forward references).
 
         NOTE: Usually get_class_decorator_hook_2 is the better option, since it
               guarantees that there are no placeholders.

--- a/mypy/plugin.py
+++ b/mypy/plugin.py
@@ -692,14 +692,33 @@ class Plugin(CommonPluginApi):
 
         The plugin can modify a TypeInfo _in place_ (for example add some generated
         methods to the symbol table). This hook is called after the class body was
-        semantically analyzed.
+        semantically analyzed, but *there may still be placeholders*.
 
-        The hook is called with full names of all class decorators, for example
+        NOTE: Usually get_class_decorator_hook_2 is the better option, since it
+              guarantees that there are no placeholders.
+
+        The hook is called with full names of all class decorators.
+
+        The hook can be called multiple times per class, so it must be
+        idempotent.
         """
         return None
 
     def get_class_decorator_hook_2(self, fullname: str
                                    ) -> Optional[Callable[[ClassDefContext], bool]]:
+        """Update class definition for given class decorators.
+
+        Similar to get_class_decorator_hook, but this runs in a later pass when
+        placeholders have been resolved.
+
+        The hook can return False if some base class hasn't been
+        processed yet using class hooks. It causes all class hooks
+        (that are run in this same pass) to be invoked another time for
+        the file(s) currently being processed.
+
+        The hook can be called multiple times per class, so it must be
+        idempotent.
+        """
         return None
 
     def get_metaclass_hook(self, fullname: str

--- a/mypy/plugin.py
+++ b/mypy/plugin.py
@@ -820,7 +820,7 @@ class ChainedPlugin(Plugin):
         return self._find_hook(lambda plugin: plugin.get_class_decorator_hook(fullname))
 
     def get_class_decorator_hook_2(self, fullname: str
-                                  ) -> Optional[Callable[[ClassDefContext], None]]:
+                                   ) -> Optional[Callable[[ClassDefContext], None]]:
         return self._find_hook(lambda plugin: plugin.get_class_decorator_hook_2(fullname))
 
     def get_metaclass_hook(self, fullname: str

--- a/mypy/plugin.py
+++ b/mypy/plugin.py
@@ -819,6 +819,10 @@ class ChainedPlugin(Plugin):
                                  ) -> Optional[Callable[[ClassDefContext], None]]:
         return self._find_hook(lambda plugin: plugin.get_class_decorator_hook(fullname))
 
+    def get_class_decorator_hook_2(self, fullname: str
+                                  ) -> Optional[Callable[[ClassDefContext], None]]:
+        return self._find_hook(lambda plugin: plugin.get_class_decorator_hook_2(fullname))
+
     def get_metaclass_hook(self, fullname: str
                            ) -> Optional[Callable[[ClassDefContext], None]]:
         return self._find_hook(lambda plugin: plugin.get_metaclass_hook(fullname))

--- a/mypy/plugin.py
+++ b/mypy/plugin.py
@@ -698,6 +698,10 @@ class Plugin(CommonPluginApi):
         """
         return None
 
+    def get_class_decorator_hook_2(self, fullname: str
+                                   ) -> Optional[Callable[[ClassDefContext], None]]:
+        return None
+
     def get_metaclass_hook(self, fullname: str
                            ) -> Optional[Callable[[ClassDefContext], None]]:
         """Update class definition for given declared metaclasses.

--- a/mypy/plugin.py
+++ b/mypy/plugin.py
@@ -699,7 +699,7 @@ class Plugin(CommonPluginApi):
         return None
 
     def get_class_decorator_hook_2(self, fullname: str
-                                   ) -> Optional[Callable[[ClassDefContext], None]]:
+                                   ) -> Optional[Callable[[ClassDefContext], bool]]:
         return None
 
     def get_metaclass_hook(self, fullname: str
@@ -820,7 +820,7 @@ class ChainedPlugin(Plugin):
         return self._find_hook(lambda plugin: plugin.get_class_decorator_hook(fullname))
 
     def get_class_decorator_hook_2(self, fullname: str
-                                   ) -> Optional[Callable[[ClassDefContext], None]]:
+                                   ) -> Optional[Callable[[ClassDefContext], bool]]:
         return self._find_hook(lambda plugin: plugin.get_class_decorator_hook_2(fullname))
 
     def get_metaclass_hook(self, fullname: str

--- a/mypy/plugins/dataclasses.py
+++ b/mypy/plugins/dataclasses.py
@@ -304,6 +304,9 @@ class DataclassTransformer:
           b: SomeOtherType = ...
 
         are collected.
+
+        Return None if some dataclass base class hasn't been processed
+        yet and thus we'll need to ask for another pass.
         """
         # First, collect attributes belonging to the current class.
         ctx = self._ctx

--- a/mypy/plugins/dataclasses.py
+++ b/mypy/plugins/dataclasses.py
@@ -110,7 +110,7 @@ class DataclassTransformer:
     def __init__(self, ctx: ClassDefContext) -> None:
         self._ctx = ctx
 
-    def transform(self) -> None:
+    def transform(self) -> bool:
         """Apply all the necessary transformations to the underlying
         dataclass so as to ensure it is fully type checked according
         to the rules in PEP 557.
@@ -119,12 +119,11 @@ class DataclassTransformer:
         info = self._ctx.cls.info
         attributes = self.collect_attributes()
         if attributes is None:
-            # Some definitions are not ready, defer() should be already called.
-            return
+            # Some definitions are not ready. we need another pass.
+            return False
         for attr in attributes:
             if attr.type is None:
-                ctx.api.defer()
-                return
+                return False
         decorator_arguments = {
             'init': _get_decorator_bool_argument(self._ctx, 'init', True),
             'eq': _get_decorator_bool_argument(self._ctx, 'eq', True),
@@ -236,6 +235,8 @@ class DataclassTransformer:
             'frozen': decorator_arguments['frozen'],
         }
 
+        return True
+
     def add_slots(self,
                   info: TypeInfo,
                   attributes: List[DataclassAttribute],
@@ -315,14 +316,11 @@ class DataclassTransformer:
 
             sym = cls.info.names.get(lhs.name)
             if sym is None:
-                # This name is likely blocked by a star import. We don't need to defer because
-                # defer() is already called by mark_incomplete().
+                # There was probably a semantic analysis error.
                 continue
 
             node = sym.node
-            if isinstance(node, PlaceholderNode):
-                # This node is not ready yet.
-                return None
+            assert not isinstance(node, PlaceholderNode)
             assert isinstance(node, Var)
 
             # x: ClassVar[int] is ignored by dataclasses.
@@ -390,6 +388,9 @@ class DataclassTransformer:
         # we'll have unmodified attrs laying around.
         all_attrs = attrs.copy()
         for info in cls.info.mro[1:-1]:
+            if 'dataclass_tag' in info.metadata and 'dataclass' not in info.metadata:
+                # We haven't processed the base class yet. Need another pass.
+                return None
             if 'dataclass' not in info.metadata:
                 continue
 
@@ -517,11 +518,21 @@ class DataclassTransformer:
         )
 
 
-def dataclass_class_maker_callback(ctx: ClassDefContext) -> None:
+def dataclass_tag_callback(ctx: ClassDefContext) -> None:
+    """Record that we have a dataclass in the main semantic analysis pass.
+
+    The later pass implemented by DataclassTransformer will use this
+    to detect dataclasses in base classes.
+    """
+    # The value is ignored, only the existence matters.
+    ctx.cls.info.metadata['dataclass_tag'] = {}
+
+
+def dataclass_class_maker_callback(ctx: ClassDefContext) -> bool:
     """Hooks into the class typechecking process to add support for dataclasses.
     """
     transformer = DataclassTransformer(ctx)
-    transformer.transform()
+    return transformer.transform()
 
 
 def _collect_field_args(expr: Expression,

--- a/mypy/plugins/dataclasses.py
+++ b/mypy/plugins/dataclasses.py
@@ -107,6 +107,15 @@ class DataclassAttribute:
 
 
 class DataclassTransformer:
+    """Implement the behavior of @dataclass.
+
+    Note that this may be executed multiple times on the same class, so
+    everything here must be idempotent.
+
+    This runs after the main semantic analysis pass, so you can assume that
+    there are no placeholders.
+    """
+
     def __init__(self, ctx: ClassDefContext) -> None:
         self._ctx = ctx
 
@@ -119,7 +128,7 @@ class DataclassTransformer:
         info = self._ctx.cls.info
         attributes = self.collect_attributes()
         if attributes is None:
-            # Some definitions are not ready. we need another pass.
+            # Some definitions are not ready. We need another pass.
             return False
         for attr in attributes:
             if attr.type is None:

--- a/mypy/plugins/default.py
+++ b/mypy/plugins/default.py
@@ -115,8 +115,6 @@ class DefaultPlugin(Plugin):
                 attrs.attr_class_maker_callback,
                 auto_attribs_default=None,
             )
-        #elif fullname in dataclasses.dataclass_makers:
-        #    return dataclasses.dataclass_class_maker_callback
         elif fullname in functools.functools_total_ordering_makers:
             return functools.functools_total_ordering_maker_callback
 

--- a/mypy/plugins/default.py
+++ b/mypy/plugins/default.py
@@ -95,7 +95,6 @@ class DefaultPlugin(Plugin):
     def get_class_decorator_hook(self, fullname: str
                                  ) -> Optional[Callable[[ClassDefContext], None]]:
         from mypy.plugins import attrs
-        from mypy.plugins import dataclasses
         from mypy.plugins import functools
 
         if fullname in attrs.attr_class_makers:
@@ -116,10 +115,19 @@ class DefaultPlugin(Plugin):
                 attrs.attr_class_maker_callback,
                 auto_attribs_default=None,
             )
-        elif fullname in dataclasses.dataclass_makers:
-            return dataclasses.dataclass_class_maker_callback
+        #elif fullname in dataclasses.dataclass_makers:
+        #    return dataclasses.dataclass_class_maker_callback
         elif fullname in functools.functools_total_ordering_makers:
             return functools.functools_total_ordering_maker_callback
+
+        return None
+
+    def get_class_decorator_hook_2(self, fullname: str
+                                   ) -> Optional[Callable[[ClassDefContext], None]]:
+        from mypy.plugins import dataclasses
+
+        if fullname in dataclasses.dataclass_makers:
+            return dataclasses.dataclass_class_maker_callback
 
         return None
 

--- a/mypy/plugins/default.py
+++ b/mypy/plugins/default.py
@@ -95,6 +95,7 @@ class DefaultPlugin(Plugin):
     def get_class_decorator_hook(self, fullname: str
                                  ) -> Optional[Callable[[ClassDefContext], None]]:
         from mypy.plugins import attrs
+        from mypy.plugins import dataclasses
         from mypy.plugins import functools
 
         if fullname in attrs.attr_class_makers:
@@ -115,13 +116,15 @@ class DefaultPlugin(Plugin):
                 attrs.attr_class_maker_callback,
                 auto_attribs_default=None,
             )
+        elif fullname in dataclasses.dataclass_makers:
+            return dataclasses.dataclass_tag_callback
         elif fullname in functools.functools_total_ordering_makers:
             return functools.functools_total_ordering_maker_callback
 
         return None
 
     def get_class_decorator_hook_2(self, fullname: str
-                                   ) -> Optional[Callable[[ClassDefContext], None]]:
+                                   ) -> Optional[Callable[[ClassDefContext], bool]]:
         from mypy.plugins import dataclasses
 
         if fullname in dataclasses.dataclass_makers:

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -1234,42 +1234,43 @@ class SemanticAnalyzer(NodeVisitor[None],
 
     def apply_class_plugin_hooks(self, defn: ClassDef) -> None:
         """Apply a plugin hook that may infer a more precise definition for a class."""
-        def get_fullname(expr: Expression) -> Optional[str]:
-            if isinstance(expr, CallExpr):
-                return get_fullname(expr.callee)
-            elif isinstance(expr, IndexExpr):
-                return get_fullname(expr.base)
-            elif isinstance(expr, RefExpr):
-                if expr.fullname:
-                    return expr.fullname
-                # If we don't have a fullname look it up. This happens because base classes are
-                # analyzed in a different manner (see exprtotype.py) and therefore those AST
-                # nodes will not have full names.
-                sym = self.lookup_type_node(expr)
-                if sym:
-                    return sym.fullname
-            return None
 
         for decorator in defn.decorators:
-            decorator_name = get_fullname(decorator)
+            decorator_name = self.get_fullname_for_hook(decorator)
             if decorator_name:
                 hook = self.plugin.get_class_decorator_hook(decorator_name)
                 if hook:
                     hook(ClassDefContext(defn, decorator, self))
 
         if defn.metaclass:
-            metaclass_name = get_fullname(defn.metaclass)
+            metaclass_name = self.get_fullname_for_hook(defn.metaclass)
             if metaclass_name:
                 hook = self.plugin.get_metaclass_hook(metaclass_name)
                 if hook:
                     hook(ClassDefContext(defn, defn.metaclass, self))
 
         for base_expr in defn.base_type_exprs:
-            base_name = get_fullname(base_expr)
+            base_name = self.get_fullname_for_hook(base_expr)
             if base_name:
                 hook = self.plugin.get_base_class_hook(base_name)
                 if hook:
                     hook(ClassDefContext(defn, base_expr, self))
+
+    def get_fullname_for_hook(self, expr: Expression) -> Optional[str]:
+        if isinstance(expr, CallExpr):
+            return self.get_fullname_for_hook(expr.callee)
+        elif isinstance(expr, IndexExpr):
+            return self.get_fullname_for_hook(expr.base)
+        elif isinstance(expr, RefExpr):
+            if expr.fullname:
+                return expr.fullname
+            # If we don't have a fullname look it up. This happens because base classes are
+            # analyzed in a different manner (see exprtotype.py) and therefore those AST
+            # nodes will not have full names.
+            sym = self.lookup_type_node(expr)
+            if sym:
+                return sym.fullname
+        return None
 
     def analyze_class_keywords(self, defn: ClassDef) -> None:
         for value in defn.keywords.values():

--- a/mypy/semanal_main.py
+++ b/mypy/semanal_main.py
@@ -82,6 +82,8 @@ def semantic_analysis_for_scc(graph: 'Graph', scc: List[str], errors: Errors) ->
     apply_semantic_analyzer_patches(patches)
     # This pass might need fallbacks calculated above.
     check_type_arguments(graph, scc, errors)
+    # Run class decorator hooks (they requite complete MROs).
+    apply_class_decorator_hooks(graph, scc, errors)
     calculate_class_properties(graph, scc, errors)
     check_blockers(graph, scc)
     # Clean-up builtins, so that TypeVar etc. are not accessible without importing.
@@ -380,6 +382,56 @@ def check_type_arguments_in_targets(targets: List[FineGrainedDeferredNode], stat
                 with errors.scope.saved_scope(saved) if errors.scope else nullcontext():
                     analyzer.recurse_into_functions = func is not None
                     target.node.accept(analyzer)
+
+
+from mypy.nodes import Expression, CallExpr, IndexExpr, RefExpr
+from mypy.plugin import ClassDefContext
+
+
+def apply_class_decorator_hooks(graph: 'Graph', scc: List[str], errors: Errors) -> None:
+    incomplete = True
+    while incomplete:
+        incomplete = False
+        for module in scc:
+            state = graph[module]
+            tree = state.tree
+            assert tree
+            for _, node, _ in tree.local_definitions():
+                if isinstance(node.node, TypeInfo):
+                    if not apply_hooks_to_class(state.manager.semantic_analyzer,
+                                                module, node.node, errors):
+                        incomplete = True
+
+
+def apply_hooks_to_class(self: SemanticAnalyzer, module: str, info: TypeInfo,
+                         errors: Errors) -> bool:
+
+    def get_fullname(expr: Expression) -> Optional[str]:
+        if isinstance(expr, CallExpr):
+            return get_fullname(expr.callee)
+        elif isinstance(expr, IndexExpr):
+            return get_fullname(expr.base)
+        elif isinstance(expr, RefExpr):
+            if expr.fullname:
+                return expr.fullname
+            # If we don't have a fullname look it up. This happens because base classes are
+            # analyzed in a different manner (see exprtotype.py) and therefore those AST
+            # nodes will not have full names.
+            sym = self.lookup_type_node(expr)
+            if sym:
+                return sym.fullname
+        return None
+
+    saved = (module, info, None)  # module, class, function
+    with errors.scope.saved_scope(saved) if errors.scope else nullcontext():
+        defn = info.defn
+        for decorator in defn.decorators:
+            decorator_name = get_fullname(decorator)
+            if decorator_name:
+                hook = self.plugin.get_class_decorator_hook_2(decorator_name)
+                if hook:
+                    hook(ClassDefContext(defn, decorator, self))
+    return True
 
 
 def calculate_class_properties(graph: 'Graph', scc: List[str], errors: Errors) -> None:

--- a/mypy/semanal_main.py
+++ b/mypy/semanal_main.py
@@ -83,8 +83,8 @@ def semantic_analysis_for_scc(graph: 'Graph', scc: List[str], errors: Errors) ->
     apply_semantic_analyzer_patches(patches)
     # This pass might need fallbacks calculated above.
     check_type_arguments(graph, scc, errors)
-    # Run class decorator hooks (they requite complete MROs).
-    apply_class_decorator_hooks(graph, scc, errors)
+    # Run class decorator hooks (they requite complete MROs and no placeholders).
+    apply_class_plugin_hooks(graph, scc, errors)
     calculate_class_properties(graph, scc, errors)
     check_blockers(graph, scc)
     # Clean-up builtins, so that TypeVar etc. are not accessible without importing.
@@ -135,7 +135,7 @@ def semantic_analysis_for_targets(
 
     check_type_arguments_in_targets(nodes, state, state.manager.errors)
     calculate_class_properties(graph, [state.id], state.manager.errors)
-    apply_class_decorator_hooks(graph, [state.id], state.manager.errors)
+    apply_class_plugin_hooks(graph, [state.id], state.manager.errors)
 
 
 def restore_saved_attrs(saved_attrs: SavedAttributes) -> None:
@@ -390,7 +390,16 @@ from mypy.nodes import Expression, CallExpr, IndexExpr, RefExpr
 from mypy.plugin import ClassDefContext
 
 
-def apply_class_decorator_hooks(graph: 'Graph', scc: List[str], errors: Errors) -> None:
+def apply_class_plugin_hooks(graph: 'Graph', scc: List[str], errors: Errors) -> None:
+    """Apply class plugin hooks within a SCC.
+
+    We run these after to the main semantic analysis so that the hooks
+    don't need to deal with incomplete definitions such as placeholder
+    types.
+
+    Note that some hooks incorrectly run during the main semantic
+    analysis pass, for historical reasons.
+    """
     incomplete = True
     while incomplete:
         incomplete = False
@@ -411,6 +420,7 @@ def apply_hooks_to_class(self: SemanticAnalyzer,
                          options: Options,
                          file_node: MypyFile,
                          errors: Errors) -> bool:
+    # TODO: Move more class-related hooks here?
     defn = info.defn
     for decorator in defn.decorators:
         with self.file_context(file_node, options, info):

--- a/mypy/semanal_main.py
+++ b/mypy/semanal_main.py
@@ -402,7 +402,7 @@ def apply_class_plugin_hooks(graph: 'Graph', scc: List[str], errors: Errors) -> 
     # If we encounter a base class that has not been processed, we'll run another
     # pass. This should eventually reach a fixed point.
     while incomplete:
-        assert num_passes < 5, "Internal error: too many class plugin hook passes"
+        assert num_passes < 10, "Internal error: too many class plugin hook passes"
         num_passes += 1
         incomplete = False
         for module in scc:

--- a/mypy/semanal_main.py
+++ b/mypy/semanal_main.py
@@ -436,7 +436,7 @@ def apply_hooks_to_class(self: SemanticAnalyzer,
             if decorator_name:
                 hook = self.plugin.get_class_decorator_hook_2(decorator_name)
                 if hook:
-                        hook(ClassDefContext(defn, decorator, self))
+                    hook(ClassDefContext(defn, decorator, self))
     return True
 
 

--- a/mypy/semanal_main.py
+++ b/mypy/semanal_main.py
@@ -397,8 +397,13 @@ def apply_class_plugin_hooks(graph: 'Graph', scc: List[str], errors: Errors) -> 
     Note that some hooks incorrectly run during the main semantic
     analysis pass, for historical reasons.
     """
+    num_passes = 0
     incomplete = True
+    # If we encounter a base class that has not been processed, we'll run another
+    # pass. This should eventually reach a fixed point.
     while incomplete:
+        assert num_passes < 5, "Internal error: too many class plugin hook passes"
+        num_passes += 1
         incomplete = False
         for module in scc:
             state = graph[module]
@@ -419,14 +424,15 @@ def apply_hooks_to_class(self: SemanticAnalyzer,
                          errors: Errors) -> bool:
     # TODO: Move more class-related hooks here?
     defn = info.defn
+    ok = True
     for decorator in defn.decorators:
         with self.file_context(file_node, options, info):
             decorator_name = self.get_fullname_for_hook(decorator)
             if decorator_name:
                 hook = self.plugin.get_class_decorator_hook_2(decorator_name)
                 if hook:
-                    hook(ClassDefContext(defn, decorator, self))
-    return True
+                    ok = ok and hook(ClassDefContext(defn, decorator, self))
+    return ok
 
 
 def calculate_class_properties(graph: 'Graph', scc: List[str], errors: Errors) -> None:

--- a/mypy/semanal_main.py
+++ b/mypy/semanal_main.py
@@ -45,6 +45,7 @@ from mypy.semanal_infer import infer_decorator_signature_if_simple
 from mypy.checker import FineGrainedDeferredNode
 from mypy.server.aststrip import SavedAttributes
 from mypy.util import is_typeshed_file
+from mypy.options import Options
 import mypy.build
 
 if TYPE_CHECKING:
@@ -399,11 +400,15 @@ def apply_class_decorator_hooks(graph: 'Graph', scc: List[str], errors: Errors) 
             for _, node, _ in tree.local_definitions():
                 if isinstance(node.node, TypeInfo):
                     if not apply_hooks_to_class(state.manager.semantic_analyzer,
-                                                module, node.node, errors):
+                                                module, node.node, state.options, tree, errors):
                         incomplete = True
 
 
-def apply_hooks_to_class(self: SemanticAnalyzer, module: str, info: TypeInfo,
+def apply_hooks_to_class(self: SemanticAnalyzer,
+                         module: str,
+                         info: TypeInfo,
+                         options: Options,
+                         file_node: MypyFile,
                          errors: Errors) -> bool:
 
     def get_fullname(expr: Expression) -> Optional[str]:
@@ -423,13 +428,13 @@ def apply_hooks_to_class(self: SemanticAnalyzer, module: str, info: TypeInfo,
         return None
 
     saved = (module, info, None)  # module, class, function
-    with errors.scope.saved_scope(saved) if errors.scope else nullcontext():
-        defn = info.defn
-        for decorator in defn.decorators:
-            decorator_name = get_fullname(decorator)
-            if decorator_name:
-                hook = self.plugin.get_class_decorator_hook_2(decorator_name)
-                if hook:
+    defn = info.defn
+    for decorator in defn.decorators:
+        decorator_name = get_fullname(decorator)
+        if decorator_name:
+            hook = self.plugin.get_class_decorator_hook_2(decorator_name)
+            if hook:
+                with self.file_context(file_node, options, info):
                     hook(ClassDefContext(defn, decorator, self))
     return True
 

--- a/mypy/semanal_main.py
+++ b/mypy/semanal_main.py
@@ -431,12 +431,12 @@ def apply_hooks_to_class(self: SemanticAnalyzer,
     saved = (module, info, None)  # module, class, function
     defn = info.defn
     for decorator in defn.decorators:
-        decorator_name = get_fullname(decorator)
-        if decorator_name:
-            hook = self.plugin.get_class_decorator_hook_2(decorator_name)
-            if hook:
-                with self.file_context(file_node, options, info):
-                    hook(ClassDefContext(defn, decorator, self))
+        with self.file_context(file_node, options, info):
+            decorator_name = get_fullname(decorator)
+            if decorator_name:
+                hook = self.plugin.get_class_decorator_hook_2(decorator_name)
+                if hook:
+                        hook(ClassDefContext(defn, decorator, self))
     return True
 
 

--- a/mypy/semanal_main.py
+++ b/mypy/semanal_main.py
@@ -135,6 +135,7 @@ def semantic_analysis_for_targets(
 
     check_type_arguments_in_targets(nodes, state, state.manager.errors)
     calculate_class_properties(graph, [state.id], state.manager.errors)
+    apply_class_decorator_hooks(graph, [state.id], state.manager.errors)
 
 
 def restore_saved_attrs(saved_attrs: SavedAttributes) -> None:

--- a/mypy/semanal_main.py
+++ b/mypy/semanal_main.py
@@ -428,7 +428,6 @@ def apply_hooks_to_class(self: SemanticAnalyzer,
                 return sym.fullname
         return None
 
-    saved = (module, info, None)  # module, class, function
     defn = info.defn
     for decorator in defn.decorators:
         with self.file_context(file_node, options, info):

--- a/mypy/semanal_main.py
+++ b/mypy/semanal_main.py
@@ -46,6 +46,7 @@ from mypy.checker import FineGrainedDeferredNode
 from mypy.server.aststrip import SavedAttributes
 from mypy.util import is_typeshed_file
 from mypy.options import Options
+from mypy.plugin import ClassDefContext
 import mypy.build
 
 if TYPE_CHECKING:
@@ -384,10 +385,6 @@ def check_type_arguments_in_targets(targets: List[FineGrainedDeferredNode], stat
                 with errors.scope.saved_scope(saved) if errors.scope else nullcontext():
                     analyzer.recurse_into_functions = func is not None
                     target.node.accept(analyzer)
-
-
-from mypy.nodes import Expression, CallExpr, IndexExpr, RefExpr
-from mypy.plugin import ClassDefContext
 
 
 def apply_class_plugin_hooks(graph: 'Graph', scc: List[str], errors: Errors) -> None:

--- a/mypy/semanal_main.py
+++ b/mypy/semanal_main.py
@@ -442,12 +442,12 @@ def apply_hooks_to_class(self: SemanticAnalyzer,
 
 def calculate_class_properties(graph: 'Graph', scc: List[str], errors: Errors) -> None:
     for module in scc:
-        tree = graph[module].tree
+        state = graph[module]
+        tree = state.tree
         assert tree
         for _, node, _ in tree.local_definitions():
             if isinstance(node.node, TypeInfo):
-                saved = (module, node.node, None)  # module, class, function
-                with errors.scope.saved_scope(saved) if errors.scope else nullcontext():
+                with state.manager.semantic_analyzer.file_context(tree, state.options, node.node):
                     calculate_class_abstract_status(node.node, tree.is_stub, errors)
                     check_protocol_status(node.node, errors)
                     calculate_class_vars(node.node)

--- a/mypy/semanal_main.py
+++ b/mypy/semanal_main.py
@@ -411,27 +411,10 @@ def apply_hooks_to_class(self: SemanticAnalyzer,
                          options: Options,
                          file_node: MypyFile,
                          errors: Errors) -> bool:
-
-    def get_fullname(expr: Expression) -> Optional[str]:
-        if isinstance(expr, CallExpr):
-            return get_fullname(expr.callee)
-        elif isinstance(expr, IndexExpr):
-            return get_fullname(expr.base)
-        elif isinstance(expr, RefExpr):
-            if expr.fullname:
-                return expr.fullname
-            # If we don't have a fullname look it up. This happens because base classes are
-            # analyzed in a different manner (see exprtotype.py) and therefore those AST
-            # nodes will not have full names.
-            sym = self.lookup_type_node(expr)
-            if sym:
-                return sym.fullname
-        return None
-
     defn = info.defn
     for decorator in defn.decorators:
         with self.file_context(file_node, options, info):
-            decorator_name = get_fullname(decorator)
+            decorator_name = self.get_fullname_for_hook(decorator)
             if decorator_name:
                 hook = self.plugin.get_class_decorator_hook_2(decorator_name)
                 if hook:

--- a/test-data/unit/check-dataclasses.test
+++ b/test-data/unit/check-dataclasses.test
@@ -1620,20 +1620,52 @@ Child(x='', y=1)
 Child(x=None, y=None)
 [builtins fixtures/dataclasses.pyi]
 
-[case testDataclassGenericInheritanceSpecialCase]
+[case testDataclassGenericInheritanceSpecialCase1]
 # flags: --python-version 3.7
 from dataclasses import dataclass
-from typing import Generic, TypeVar
+from typing import Generic, TypeVar, List
 
 T = TypeVar("T")
 
 @dataclass
 class Parent(Generic[T]):
-    key: str
+    x: List[T]
 
 @dataclass
 class Child1(Parent["Child2"]): ...
 
 @dataclass
 class Child2(Parent["Child1"]): ...
+
+def f(c: Child2) -> None:
+    reveal_type(Child1([c]).x)  # N: Revealed type is "builtins.list[__main__.Child2]"
+
+def g(c: Child1) -> None:
+    reveal_type(Child2([c]).x)  # N: Revealed type is "builtins.list[__main__.Child1]"
+[builtins fixtures/dataclasses.pyi]
+
+[case testDataclassGenericInheritanceSpecialCase2]
+# flags: --python-version 3.7
+from dataclasses import dataclass
+from typing import Generic, TypeVar
+
+T = TypeVar("T")
+
+# A subclass might be analyzed before base in import cycles.  They are
+# defined here in reversed order to simulate this.
+
+@dataclass
+class Child1(Parent["Child2"]):
+    x: int
+
+@dataclass
+class Child2(Parent["Child1"]):
+    y: int
+
+@dataclass
+class Parent(Generic[T]):
+    key: str
+
+Child1(x=1, key='')
+Child2(y=1, key='')
 [builtins fixtures/dataclasses.pyi]

--- a/test-data/unit/check-dataclasses.test
+++ b/test-data/unit/check-dataclasses.test
@@ -1619,3 +1619,21 @@ Child(x='', y='')  # E: Argument "y" to "Child" has incompatible type "str"; exp
 Child(x='', y=1)
 Child(x=None, y=None)
 [builtins fixtures/dataclasses.pyi]
+
+[case testDataclassGenericInheritanceSpecialCase]
+# flags: --python-version 3.7
+from dataclasses import dataclass
+from typing import Generic, TypeVar
+
+T = TypeVar("T")
+
+@dataclass
+class Parent(Generic[T]):
+    key: str
+
+@dataclass
+class Child1(Parent["Child2"]): ...
+
+@dataclass
+class Child2(Parent["Child1"]): ...
+[builtins fixtures/dataclasses.pyi]

--- a/test-data/unit/check-dataclasses.test
+++ b/test-data/unit/check-dataclasses.test
@@ -1669,3 +1669,19 @@ class Parent(Generic[T]):
 Child1(x=1, key='')
 Child2(y=1, key='')
 [builtins fixtures/dataclasses.pyi]
+
+[case testDataclassGenericWithBound]
+# flags: --python-version 3.7
+from dataclasses import dataclass
+from typing import Generic, TypeVar
+
+T = TypeVar("T", bound="C")
+
+@dataclass
+class C(Generic[T]):
+    x: int
+
+c: C[C]
+d: C[str]  # E: Type argument "str" of "C" must be a subtype of "C[Any]"
+C(x=2)
+[builtins fixtures/dataclasses.pyi]

--- a/test-data/unit/check-dataclasses.test
+++ b/test-data/unit/check-dataclasses.test
@@ -1207,7 +1207,7 @@ class A2:
 from dataclasses import dataclass
 
 @dataclass
-class A:  # E: Name "x" already defined (possibly by an import)
+class A:
     x: int = 0
     x: int = 0  # E: Name "x" already defined on line 7
 

